File: .github/scripts/seed_release_bump.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import re
import sys
from pathlib import Path


SEED_RELEASE_RE = re.compile(r'^(seed_release:\s*)"?([^"#\s]+)"?(.*)$')


def bump_minor(version: str) -> str:
    parts = version.strip().split(".")
    if len(parts) != 3 or not all(part.isdigit() for part in parts):
        raise ValueError(f"seed_release is not a major.minor.patch version: {version}")
    major, minor, _patch = (int(part) for part in parts)
    return f"{major}.{minor + 1}.0"


def read_seed_release(path: Path) -> str:
    for line in path.read_text(encoding="utf-8").splitlines():
        match = SEED_RELEASE_RE.match(line)
        if match:
            return match.group(2)
    raise ValueError(f"{path} does not contain seed_release")


def bump_file(path: Path) -> tuple[str, str]:
    lines = path.read_text(encoding="utf-8").splitlines(keepends=True)
    current = ""
    next_version = ""
    out: list[str] = []
    replaced = False
    for line in lines:
        newline = "\n" if line.endswith("\n") else ""
        body = line[:-1] if newline else line
        match = SEED_RELEASE_RE.match(body)
        if match and not replaced:
            current = match.group(2)
            next_version = bump_minor(current)
            out.append(f'{match.group(1)}"{next_version}"{match.group(3)}{newline}')
            replaced = True
        else:
            out.append(line)
    if not replaced:
        raise ValueError(f"{path} does not contain seed_release")
    path.write_text("".join(out), encoding="utf-8")
    return current, next_version


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser()
    sub = parser.add_subparsers(dest="command", required=True)

    next_minor = sub.add_parser("next-minor")
    next_minor.add_argument("version")

    read = sub.add_parser("read")
    read.add_argument("path", type=Path)

    bump = sub.add_parser("bump-file")
    bump.add_argument("path", type=Path)

    args = parser.parse_args(argv[1:])
    try:
        if args.command == "next-minor":
            print(bump_minor(args.version))
        elif args.command == "read":
            print(read_seed_release(args.path))
        elif args.command == "bump-file":
            current, next_version = bump_file(args.path)
            print(f"current={current}")
            print(f"next={next_version}")
        else:
            raise AssertionError(args.command)
    except ValueError as exc:
        print(str(exc), file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))

File: .github/scripts/seed_release_bump.go
//go:build ignore

package main

import (
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

var seedReleaseRE = regexp.MustCompile(`^(seed_release:\s*)"?([^"#\s]+)"?(.*)$`)

func bumpPatch(version string) (string, error) {
	parts := strings.Split(strings.TrimSpace(version), ".")
	if len(parts) != 3 {
		return "", fmt.Errorf("seed_release is not a major.minor.patch version: %s", version)
	}
	values := make([]int, 3)
	for i, part := range parts {
		if part == "" || strings.Trim(part, "0123456789") != "" {
			return "", fmt.Errorf("seed_release is not a major.minor.patch version: %s", version)
		}
		value, err := strconv.Atoi(part)
		if err != nil {
			return "", fmt.Errorf("seed_release is not a major.minor.patch version: %s", version)
		}
		values[i] = value
	}
	return fmt.Sprintf("%d.%d.%d", values[0], values[1], values[2]+1), nil
}

func readSeedRelease(path string) (string, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	for _, line := range strings.Split(string(data), "\n") {
		match := seedReleaseRE.FindStringSubmatch(strings.TrimSuffix(line, "\r"))
		if match != nil {
			return match[2], nil
		}
	}
	return "", fmt.Errorf("%s does not contain seed_release", path)
}

func bumpFile(path string) (string, string, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return "", "", err
	}
	lines := strings.SplitAfter(string(data), "\n")
	out := make([]string, 0, len(lines))
	current := ""
	next := ""
	replaced := false
	for _, line := range lines {
		if line == "" {
			continue
		}
		newline := ""
		body := line
		if strings.HasSuffix(body, "\n") {
			newline = "\n"
			body = strings.TrimSuffix(body, "\n")
		}
		match := seedReleaseRE.FindStringSubmatch(strings.TrimSuffix(body, "\r"))
		if match != nil && !replaced {
			current = match[2]
			next, err = bumpPatch(current)
			if err != nil {
				return "", "", err
			}
			out = append(out, fmt.Sprintf(`%s"%s"%s%s`, match[1], next, match[3], newline))
			replaced = true
			continue
		}
		out = append(out, line)
	}
	if !replaced {
		return "", "", fmt.Errorf("%s does not contain seed_release", path)
	}
	if err := os.WriteFile(path, []byte(strings.Join(out, "")), 0o644); err != nil {
		return "", "", err
	}
	return current, next, nil
}

func setFile(path, version string) (string, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	lines := strings.SplitAfter(string(data), "\n")
	out := make([]string, 0, len(lines))
	current := ""
	replaced := false
	for _, line := range lines {
		if line == "" {
			continue
		}
		newline := ""
		body := line
		if strings.HasSuffix(body, "\n") {
			newline = "\n"
			body = strings.TrimSuffix(body, "\n")
		}
		match := seedReleaseRE.FindStringSubmatch(strings.TrimSuffix(body, "\r"))
		if match != nil && !replaced {
			current = match[2]
			out = append(out, fmt.Sprintf(`%s"%s"%s%s`, match[1], version, match[3], newline))
			replaced = true
			continue
		}
		out = append(out, line)
	}
	if !replaced {
		return "", fmt.Errorf("%s does not contain seed_release", path)
	}
	if err := os.WriteFile(path, []byte(strings.Join(out, "")), 0o644); err != nil {
		return "", err
	}
	return current, nil
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "usage: seed_release_bump <next-patch|read|bump-file|set-file> [...]")
		os.Exit(2)
	}
	var err error
	switch os.Args[1] {
	case "next-patch":
		if len(os.Args) != 3 {
			err = fmt.Errorf("usage: seed_release_bump next-patch <version>")
			break
		}
		var next string
		next, err = bumpPatch(os.Args[2])
		if err == nil {
			fmt.Println(next)
		}
	case "read":
		if len(os.Args) != 3 {
			err = fmt.Errorf("usage: seed_release_bump read <path>")
			break
		}
		var version string
		version, err = readSeedRelease(os.Args[2])
		if err == nil {
			fmt.Println(version)
		}
	case "bump-file":
		if len(os.Args) != 3 {
			err = fmt.Errorf("usage: seed_release_bump bump-file <path>")
			break
		}
		var current, next string
		current, next, err = bumpFile(os.Args[2])
		if err == nil {
			fmt.Printf("current=%s\nnext=%s\n", current, next)
		}
	case "set-file":
		if len(os.Args) != 4 {
			err = fmt.Errorf("usage: seed_release_bump set-file <path> <version>")
			break
		}
		var current string
		current, err = setFile(os.Args[2], os.Args[3])
		if err == nil {
			fmt.Printf("current=%s\nnext=%s\n", current, os.Args[3])
		}
	default:
		err = fmt.Errorf("unknown command: %s", os.Args[1])
	}
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}
